File: leetcode/string/345.py
class Solution:

    def reverseVowels(self, s: str) -> str:
        '''反转该字符串中的元音字母
        
        @Note:
            类似快排，左右指针
        '''
        vowels = {'a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'}
        i, j = 0, len(s) - 1
        s = list(s)
        while i < j:
            while i < j and s[i] not in vowels:
                i += 1
            while i < j and s[j] not in vowels:
                j -= 1
            if i < j:
                s[i], s[j] = s[j], s[i]
            else:
                break
        return ''.join(s)

File: leetcode/string/345.cpp
#include <iostream>
#include <set>

class Solution {
 public:
  std::string reverse_vowels(std::string& str) {
    std::set<char> vowels = {'a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'};
    size_t l = 0, r = str.size() - 1;
    while (l < r) {
      while (l < r && vowels.find(str[l]) == vowels.end()) {
        ++l;
      }
      while (l < r && vowels.find(str[r]) == vowels.end()) {
        --r;
      }
      if (l < r) {
        std::swap(str[l], str[r]);
        ++l;
        --r;
      } else {
        break;
      }
    }
    return str;
  }
};

int main() {
  std::string str = "leetcode";

  Solution solution = Solution();
  std::string reverse = solution.reverse_vowels(str);

  std::cout << "reverse: " << reverse << std::endl;
}